Plugins/MultiplayerSessionPlugin/Source/MultiplayerSessionPlugin/Private/Menu.cpp

```diff
@@ -51,34 +70,135 @@ void UMenu::OnLevelRemovedFromWorld(ULevel* InLevel, UWorld* InWorld)
 {
 	MenuTeardown();
 	Super::OnLevelRemovedFromWorld(InLevel, InWorld);
-	
+
 }
 
-void UMenu::onHostButtonClicked()
-{		
-	
+void UMenu::OnHostButtonClicked()
+{
+	if (MultiplayerSessionSubsystem) {
+		MultiplayerSessionSubsystem->CreateSession(NumPublicConnections, MatchType);
+	}
+}
+
+void UMenu::OnJoinButtonClicked()
+{
 	if (GEngine)
 	{
 		GEngine->AddOnScreenDebugMessage(
 			-1,
 			15.f,
 			FColor::Yellow,
-			FString(TEXT("Host Button Clicked"))
+			FString(TEXT("Join Button Clicked"))
 		);
 	}
+	if (MultiplayerSessionSubsystem) {
+		MultiplayerSessionSubsystem->FindSessions(20);
+	}
 }
 
-void UMenu::onJoinButtonClicked()
+void UMenu::OnCreateSession(bool bWasSuccessful)
 {
 	if (GEngine)
 	{
 		GEngine->AddOnScreenDebugMessage(
 			-1,
 			15.f,
 			FColor::Yellow,
-			FString(TEXT("Join Button Clicked"))
+			FString(TEXT("onCreateSession called"))
 		);
 	}
+	if (bWasSuccessful)
+	{
+		UWorld* World = GetWorld();
+		if (World)
+		{
+			World->ServerTravel("/Game/ThirdPerson/Maps/Lobby?listen");
+
+			if (GEngine)
+			{
+				GEngine->AddOnScreenDebugMessage(
+					-1,
+					15.f,
+					FColor::Yellow,
+					FString(TEXT("Session Created Successfully"))
+				);
+			}
+		}
+	}
+	else
+	{
+		if (GEngine)
+		{
+			GEngine->AddOnScreenDebugMessage(
+				-1,
+				15.f,
+				FColor::Red,
+				FString(TEXT("Session Creation Failed"))
+			);
+		}
+	}
+}
+
+void UMenu::OnFindSession(const TArray<FOnlineSessionSearchResult>& SearchResults, bool bWasSuccessful)
+{
+
+	if (MultiplayerSessionSubsystem == nullptr)
+	{
+		return;
+	}
+	
+	// select most appropriate match 
+	for (auto Result : SearchResults)
+	{
+		FString Id = Result.GetSessionIdStr();
+		FString User = Result.Session.OwningUserName;
+		FString SettingsMatchType;
+		Result.Session.SessionSettings.Get(FName("MatchType"), SettingsMatchType);
+		
+		if (GEngine)
+		{
+			GEngine->AddOnScreenDebugMessage(
+				-1,
+				15.f,
+				FColor::Cyan,
+				FString::Printf(TEXT("Id: %s, User: %s"), *Id, *User)
+			);
+		}
+		if (SettingsMatchType == MatchType) 
+		{
+			MultiplayerSessionSubsystem->JoinSession(Result);
+			return;
+		}
+		
+	}
+}
+
+void UMenu::OnJoinSession(EOnJoinSessionCompleteResult::Type Result)
+{
+	IOnlineSubsystem* Subsystem = IOnlineSubsystem::Get();
```

Comment:
don't love bringing the subsystem in, this could be refactored later. We're not creating a codependency which is the key rule we don't want to break here, but still- this could be cleaner